pallets/fusion/src/lib.rs

```diff
@@ -0,0 +1,2290 @@
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// A sudo-able call.
+		type RuntimeCall: Parameter
+			+ UnfilteredDispatchable<RuntimeOrigin = Self::RuntimeOrigin>
+			+ GetDispatchInfo;
+
+		/// Type representing the weight of this pallet
+		type WeightInfo: WeightInfo;
+
+		/// Currency type for this pallet.
+		type Currency: LockableCurrency<Self::AccountId, Moment = BlockNumberFor<Self>>;
+
+		/// The destination for rewards that were not claimed after 'HistoryDepth' eras.
+		type RewardRemainder: OnUnbalanced<NegativeImbalanceOf<Self>>;
+
+		/// Pallet id used to derive accounts used by the pallet
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Maximum allowed for the currency name
+		#[pallet::constant]
+		type MaxCurrencyName: Get<u32>;
+
+		/// Maximum number of members in a pool
+		#[pallet::constant]
+		type MaxMembersPerPool: Get<u32>;
+
+		/// Maximum number of selectable targets for a pool
+		#[pallet::constant]
+		type MaxTargets: Get<u32>;
+
+		/// Maximum number of parallel partial unbonds
+		#[pallet::constant]
+		type MaxUnbonding: Get<u32>;
+
+		/// Maximum number of parallel slashes
+		#[pallet::constant]
+		type MaxSlashes: Get<u32>;
+
+		/// Period for funds to be available after unbonding
+		#[pallet::constant]
+		type BondingDuration: Get<EraIndex>;
+
+		/// Number of era for which to keep Fusion data
+		#[pallet::constant]
+		type HistoryDepth: Get<u32>;
+
+		/// A provider that gives the current era.
+		type EraProvider: EraProvider;
+
+		/// Number of eras that slashes are deferred by, after computation.
+		///
+		/// This should be less than the bonding duration. Set to 0 if slashes
+		/// should be applied immediately, without opportunity for intervention.
+		#[pallet::constant]
+		type SlashDeferDuration: Get<EraIndex>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn fusion_paused)]
+	/// A storage to track whether the Fusion pallet is paused.
+	pub type FusionPaused<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+	/// Allow access to Fusion accounts on chain
+	/// Cannot be changed
+	#[pallet::storage]
+	#[pallet::getter(fn fusion_accounts)]
+	pub type FusionAccounts<T: Config> = StorageValue<_, PalletAccounts<T>, ValueQuery>;
+
+	/// Stores all the fusion currencies
+	#[pallet::storage]
+	#[pallet::getter(fn fusion_currencies)]
+	pub type FusionCurrencies<T: Config> =
+		StorageMap<_, Blake2_128Concat, CurrencyId, FusionCurrency<T>, OptionQuery>;
+
+	/// Stores the number of currencies created
+	#[pallet::storage]
+	#[pallet::getter(fn currency_count)]
+	pub type FusionCurrencyCount<T> = StorageValue<_, u32, ValueQuery>;
+
+	/// Stores all the fusion pools
+	#[pallet::storage]
+	#[pallet::getter(fn fusion_pools)]
+	pub type FusionPools<T: Config> =
+		StorageMap<_, Blake2_128Concat, PoolId, FusionPool<T>, OptionQuery>;
```

Comment:
We can use a non-cryptographic hasher here. This applies to many other storage as well